README.pod
=encoding utf8


This is Perl module B<App::aki>.

=begin html

<a href="http://travis-ci.org/bayashi/App-aki"><img src="https://secure.travis-ci.org/bayashi/App-aki.png"/></a> <a href="https://coveralls.io/r/bayashi/App-aki"><img src="https://coveralls.io/repos/bayashi/App-aki/badge.png?branch=master"/></a>

=end html


=head1 INSTALLATION

App::aki installation is straightforward. If your CPAN shell is set up,
you should just be able to do

    % cpan App::aki

Download it, unpack it, then build it as per the usual:

    % perl Makefile.PL
    % make && make test

Then install it:

    % make install


=head1 DOCUMENTATION

App::aki documentation is available as in POD. So you can do:

    % perldoc App::aki


=head1 REPOSITORY

App::aki is hosted on github
L<http://github.com/bayashi/App-aki>


=head1 LICENSE

This module is free software; you can redistribute it and/or
modify it under the same terms as Perl itself. See L<perlartistic>.


=head1 AUTHOR

Dai Okabayashi E<lt>bayashi@cpan.orgE<gt>


Changes
Revision history for Perl extension App::aki

0.12    Fri Jul 31 09:34:21 JST 2015
        - mock HTTP::Request to stop real request

0.11    Wed Jul 29 21:40:58 JST 2015
        - support Data:Printer 0.36

0.10    Mon Jul 15 00:29:40 JST 2013
        - '--rc' option has been treated by Config::CmdRC

0.09    Sun Jul 14 10:17:29 JST 2013
        - POD enhancement
        - tiny fixes

0.08    Sat Jul 13 23:33:26 JST 2013
        - NEW OPTION: '--referer' for request header

0.07    Sat Jul 13 22:47:19 JST 2013
        - rcfile has been supported by Config::CmdRC
        - NEW OPTIONS:
            - '--cookie' for request with cookie
            - '--cookie-jar' for saving response cookie
            - '--user' for BASIC authentication
            - '--header' for adding a custom http request header

0.06    Sun Jun  9 00:35:57 JST 2013
        - fixed warnings in perl 5.12 or later

0.05    Sun Jun  9 00:16:39 JST 2013
        - added new options(print_escapes, stderr, indent)

0.04    Sat Jun  1 23:30:09 JST 2013
        - new feature: --rc option, seeL<aki#CONFIGURATION>
        - new feature: be able to send request with cookie, --cookie=cookie/file/apth
        - changed a way how to detect json.
        - tiny fixes(wrong POD etc..)

0.03    Sat Jun  1 18:58:00 JST 2013
        - added encoding option: in-enc/out-enc
        - supported unicode content correctly
        - call env_proxy before request
        - tiny pod enhancement
        - refactor

0.02    Sat Jun  1 14:57:41 JST 2013
        - specify aki command in Build.PL

0.01    Sat Jun  1 01:03:58 JST 2013
        - original version


t/akiUtil.pm
package t::akiUtil;
use strict;
use warnings;
use Test::Mock::LWP;
use Capture::Tiny qw/capture/;
use App::aki;
use parent qw/Exporter/;
our @EXPORT_OK = qw/
    result
/;

sub result {
    my ($cmd_args, $mock_args) = @_;

    unshift @{$mock_args}, ['ua', 'request' => sub { HTTP::Response->new } ];

    for my $r (@{$mock_args}) {
        my $mock = shift @{$r};
        my $mock_obj = ($mock eq 'res') ? $Mock_response
                     : ($mock eq 'req') ? $Mock_request  : $Mock_ua;
        $mock_obj->mock(@{$r});
    }

    my ($stdout, $stderr, @result) = capture {
        App::aki->run(@{$cmd_args});
    };

    return($stdout, $stderr, @result);
}

1;


lib/App/aki.pm
package App::aki;
use strict;
use warnings;
use Getopt::Long qw/GetOptionsFromArray/;
use LWP::UserAgent;
use HTTP::Request;
use HTTP::Cookies;
use Data::Printer qw//;
use Encode qw//;
use File::Spec;
use Config::CmdRC '.akirc';

our $VERSION = '0.12';

# Every decode routine MUST return the UNICODE string.
our %DECODERS = (
    json => +{
        class  => 'JSON',
        detect => sub {
            my $res = shift;
            my $ct = $res->content_type;
            return 1 if $ct =~ m!json!i;
        },
        decode => sub {
            my $content = shift;
            my $json = JSON->new->utf8;
            $json->decode($content);
        },
    },
    xml => +{
        class  => 'XML::TreePP',
        detect => sub {
            my $res = shift;
            my $ct = $res->content_type;
            return 1 if $ct =~ m!xml!i;
        },
        decode => sub {
            my $content = shift;
            my $xml = XML::TreePP->new(utf8_flag => 1);
            $xml->parse($content);
        },
    },
    yaml => +{
        class  => 'YAML::Syck',
        detect => sub {
            my $res = shift;
            my $ct = $res->content_type;
            return 1 if $ct =~ m!yaml!i;
        },
        decode => sub {
            my $content = shift;
            $YAML::Syck::ImplicitUnicode = 1;
            YAML::Syck::Load($content);
        },
    },
    messagepack => +{
        class  => 'Data::MessagePack',
        detect => sub {
            my $res = shift;
            my $ct = $res->content_type;
            return 1 if $ct =~ m!msgpack!i;
        },
        decode => sub {
            my $content = shift;
            my $mp = Data::MessagePack->new->utf8;
            $mp->decode($content);
        },
    },
);

sub run {
    my $self = shift;
    my @argv = @_;

    my $config = RC();
    _merge_opt($config, @argv);

    my $res = _request($config);

    if ($config->{cookie_jar}) {
        my $cookie_jar = HTTP::Cookies->new;
        $cookie_jar->extract_cookies($res);
        $cookie_jar->save($config->{cookie_jar});
    }

    if ($config->{raw}) {
        print $res->content;
        exit;
    }

    my $decoded = _decode($config, $res);
    my $dump    = _dumper($config, $decoded);

    my $output = Encode::encode($config->{out_enc}, "---\n$dump\n---\n");
    if ($config->{stderr}) {
        print STDERR $output;
    }
    else {
        print STDOUT $output;
    }
}

sub _dumper {
    my ($config, $hash) = @_;

    my $dump = Data::Printer::p(
        $hash,
        return_value  => 'dump',
        colored       => $config->{color},
        index         => 0,
        print_escapes => $config->{print_escapes},
        indent        => $config->{indent},
    );
    $dump =~ s!^[^\n]+\n!!;
    $dump =~ s![\r\n]}$!!;

    return $dump;
}

sub _decode {
    my ($config, $res) = @_;

    my $decoded = _decoder($config, $res);

    if ($decoded && $config->{pointer}) {
        require JSON::Pointer;
        $decoded = JSON::Pointer->get($decoded, $config->{pointer});
    }

    unless ($decoded) {
        _error("could not decode the content.");
    }

    return $decoded;
}

sub _decoder {
    my ($config, $res) = @_;

    my $decoded;
    if ( my $decoder = $DECODERS{ lc($config->{decoder} || '') } ) {
        $decoded = _decoding($config, $decoder, $res);
    }
    else {
        for my $name (keys %DECODERS) {
            my $decoder = $DECODERS{$name};
            next unless $decoder->{detect}->($res);
            $decoded = _decoding($config, $decoder, $res);
            last;
        }
    }
    return $decoded;
}

sub _decoding {
    my ($config, $decoder, $res) = @_;

    _load_class( _class2path($decoder->{class}) );
    _show_verbose('decode class', $decoder->{class}) if $config->{verbose};
    my $content = $res->content;
    if ($config->{in_enc} !~ m!^utf\-?8$!i) {
        Encode::from_to($content, $config->{in_enc} => 'utf8');
    }
    return $decoder->{decode}->($content);
}

sub _error {
    my $msg = shift;

    warn "ERROR: $msg\n";
    exit;
}

sub _load_class {
    my $path = shift;

    eval {
        require $path;
        $path->import;
    };
    die $@ if $@;
}

sub _class2path {
    my $class = shift;

    $class  =~ s!::!/!g;
    $class .= '.pm';

    return $class;
}

sub _request {
    my $config = shift;

    my ($ua, $req) = _prepare_request($config);

    if ($config->{verbose}) {
        _show_verbose('request', $req->as_string);
        _show_verbose('headers', $ua->default_headers->as_string);
    }

    my $res = $ua->request($req);
    if ($res->is_success) {
        if ($config->{verbose}) {
            _show_verbose('response', $res->status_line);
            _show_verbose('response content_type', $res->header('Content_Type'));
            _show_verbose('response content length', length $res->content);
        }
        return $res;
    }
    else {
        die $res->status_line;
    }
}

sub _prepare_request {
    my $config = shift;

    my $ua = LWP::UserAgent->new(
        agent   => $config->{agent} || __PACKAGE__. "/$VERSION",
        timeout => $config->{timeout},
    );
    if ($config->{header} && ref $config->{header} eq 'ARRAY') {
        for my $h (@{$config->{header}}) {
            my ($field, $value) = split /:\s+?/, $h;
            $ua->default_header($field => $value);
        }
    }
    if ($config->{referer}) {
        $ua->default_header(referer => $config->{referer});
    }
    $ua->env_proxy;
    my $req = HTTP::Request->new(
        uc($config->{method}) => $config->{url},
    );

    if ($config->{user}) {
        my ($user, $passwd) = split /:/, $config->{user};
        $req->authorization_basic($user, $passwd);
    }

    if ($config->{cookie}) {
        $ua->cookie_jar({ file => $config->{cookie} });
    }

    return($ua, $req);
}

sub _show_verbose {
    my ($label, $line) = @_;

    $line =~ s![\r\n]+$!!;
    print "[$label]\n$line\n";
}

sub _merge_opt {
    my ($config, @argv) = @_;

    Getopt::Long::Configure('bundling');
    GetOptionsFromArray(
        \@argv,
        'd|decoder=s'   => \$config->{decoder},
        'm|method=s'    => \$config->{method},
        'timeout=i'     => \$config->{timeout},
        'H|header=s@'   => \$config->{header},
        'e|referer=s'   => \$config->{referer},
        'b|cookie=s'    => \$config->{cookie},
        'c|cookie-jar=s' => \$config->{cookie_jar},
        'u|user=s'      => \$config->{user},
        'p|pointer=s'   => \$config->{pointer},
        'ie|in-enc=s'   => \$config->{in_enc},
        'oe|out-enc=s'  => \$config->{out_enc},
        'agent=s'       => \$config->{agent},
        'color'         => \$config->{color},
        'print_escapes' => \$config->{print_escapes},
        'stderr'        => \$config->{stderr},
        'indent=i'      => \$config->{indent},
        'raw'           => \$config->{raw},
        'verbose'       => \$config->{verbose},
        'rc=s'          => \$config->{rc},
        'h|help'        => sub {
            _show_usage(1);
        },
        'v|version'   => sub {
            print "aki $VERSION\n";
            exit 1;
        },
    ) or _show_usage(2);

    $config->{agent}   ||= "aki/$VERSION";
    $config->{method}  ||= 'GET';
    $config->{timeout} ||= 10;
    $config->{color}   ||= 0;

    $config->{out_enc} ||= 'utf8';
    $config->{in_enc}  ||= 'utf8';

    $config->{indent}  ||= 4;

    $config->{url} = shift @argv;
}

sub _show_usage {
    my $exitval = shift;

    require Pod::Usage;
    Pod::Usage::pod2usage($exitval);
}

1;

__END__

=head1 NAME

App::aki - The command-line data processor for web content


=head1 SYNOPSIS

    $ aki http://example.com/json --pointer '/foo/bar/0'
    ---
        baz     123
        qux     "foobar"
    ---


=head1 DESCRIPTION

C<App::aki> is the module for data processor in the command-line.

The L<aki> command sends a request to any endpoint. Then the response is automatically detected and shown up as the dump of human readable result. Moreover, '--pointer' option can specify values what you want.

See: L<aki> command to know more details.

=head1 METHODS

=head2 run

execute process


=head1 REPOSITORY

App::aki is hosted on github
<http://github.com/bayashi/App-aki>


=head1 AUTHOR

Dai Okabayashi E<lt>bayashi@cpan.orgE<gt>


=head1 SEE ALSO

L<aki>

C<aki> was inspired from these tools.

curl:<http://curl.haxx.se/>, jq:<http://stedolan.github.io/jq/>.


=head1 LICENSE

This module is free software; you can redistribute it and/or
modify it under the same terms as Perl itself. See L<perlartistic>.

=cut


Build.PL
use strict;
use warnings;
use Module::Build;

my $builder = Module::Build->new(
    name                => 'App-aki',
    module_name         => 'App::aki',
    license             => 'perl',
    dist_author         => 'Dai Okabayashi <bayashi@cpan.org>',
    dist_version_from   => 'lib/App/aki.pm',
    configure_requires  => {
        'Module::Build' => 0.38,
    },
    build_requires      => {
        'Test::More' => 0.88,
        'Test::Mock::LWP' => 0,
        'Capture::Tiny' => 0,
        'Exporter' => 0,
        'Test::Fake::HTTPD' => 0,
        'HTTP::Response' => 0,
        'File::Temp' => 0,
    },
    requires => {
        'perl' => '5.008005',
        'Getopt::Long' => 0,
        'Pod::Usage' => 0,
        'LWP::UserAgent' => 0,
        'HTTP::Request' => 0,
        'HTTP::Cookies' => 0,
        'Data::Printer' => 0.36,
        'JSON::Pointer' => 0,
        'JSON' => 0,
        'XML::TreePP' => 0,
        'YAML::Syck' => 0,
        'Data::MessagePack' => 0,
        'Encode' => 0,
        'File::Spec' => 0,
        'Config::CmdRC' => 0.06,
    },
    script_files        => [ 'script/aki' ],
    add_to_cleanup      => [ 'App-aki-*' ],
    meta_merge     => {
      resources => {
        repository  => 'http://github.com/bayashi/App-aki',
        bugtracker  => 'http://github.com/bayashi/App-aki/issues',
      }
    },
    test_files => (-d '.git' || $ENV{RELEASE_TESTING}) ? ['t/', 'xt/'] : ['t/'],
    recursive_test_files => 1,
);

$builder->create_build_script();
